spring-cloud-stream/src/main/java/org/springframework/cloud/stream/function/StreamBridgeUtils.java

```diff
@@ -0,0 +1,140 @@
+ * While in a common scenario of a typical spring-cloud-stream application user rarely
+ * has to manually send data, there are times when the sources of data are outside of
+ * spring-cloud-stream context and therefore we need to bridge such foreign sources
+ * with spring-cloud-stream.
+ * <br><br>
+ * This utility class allows user to do just that - <i>bridge non-spring-cloud-stream applications
+ * with spring-cloud-stream</i> by providing a mechanism (bridge) to send data to an output binding while
+ * maintaining the  same invocation contract (i.e., type conversion, partitioning etc) as if it was
+ * done through a declared function.
+ *
+ * @author Oleg Zhurakousky
+ * @since 3.0.3
+ *
+ */
+public final class StreamBridgeUtils implements SmartInitializingSingleton {
+
+	private final Map<String, DirectWithAttributesChannel> outputChannelsOnly = new HashMap<>();
+
+	private final FunctionCatalog functionCatalog;
+
+	private final FunctionRegistry functionRegistry;
+
+	private BindingServiceProperties bindingServiceProperties;
+
+	private ConfigurableApplicationContext applicationContext;
+
+	private boolean initialized;
+
+	/**
+	 *
+	 * @param functionCatalog instance of {@link FunctionCatalog}
+	 * @param functionRegistry instance of {@link FunctionRegistry}
+	 * @param bindingServiceProperties instance of {@link BindingServiceProperties}
+	 * @param applicationContext instance of {@link ConfigurableApplicationContext}
+	 */
+	StreamBridgeUtils(FunctionCatalog functionCatalog, FunctionRegistry functionRegistry,
+			BindingServiceProperties bindingServiceProperties, ConfigurableApplicationContext applicationContext) {
+		this.functionCatalog = functionCatalog;
+		this.functionRegistry = functionRegistry;
+		this.applicationContext = applicationContext;
+		this.bindingServiceProperties = bindingServiceProperties;
+	}
+
+	/**
+	 * Sends 'data' to an output binding specified by 'bindingName' argument while
+	 * using default content type to deal with output type conversion (if necessary).
+	 * @param bindingName the name of the output binding
+	 * @param data the data to send
+	 * @return true if data was sent successfully, otherwise false or throws an exception.
+	 */
+	public boolean send(String bindingName, Object data) {
+		return this.send(bindingName, data, MimeTypeUtils.APPLICATION_JSON);
+	}
+
+	/**
+	 * Sends 'data' to an output binding specified by 'bindingName' argument while
+	 * using the content type specified by the 'outputContentType' argument to deal
+	 * with output type conversion (if necessary).
+	 * @param bindingName the name of the output binding
+	 * @param data the data to send
+	 * @param outputContentType content type to be used to deal with output type conversion
+	 * @return true if data was sent successfully, otherwise false or throws an exception.
+	 */
+	@SuppressWarnings("unchecked")
+	public boolean send(String bindingName, Object data, MimeType outputContentType) {
```

Comment:
If the mime type is not the default one, then are we asking applications to specify the `MimeType`? Is that the reason why this is public?

Reply:
Yes, basically this is feature of s-c-function. If there is no output content type then no conversion will happen at all, so if you look at FunctionConfiguration we always execute lookup version with output CT to ensure that the outgoing message is converted back to `Message<byte[]>`

But for this case it's even more important since this would allow user skip the creation fo the entire message just to provide custom CT header.